### PR TITLE
[Pallas] Support segment ids in flash attention

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This PR is to add segment ids to the flash attention wrapper. The segment ids are a way to create an attention mask where each token can only attend to other tokens within the same segment. The mask is therefore a block diagonal matrix.

To support it, we further split the flash attention forward into tracing and execution part, and implement all the shape operations to make it compatible with the kernel.

Test Plan:
PJRT_DEVICE=TPU python test/test_pallas.py

### CODE REVIEW

**Comment by JackCaoG** [comment]
Is this ready for review?

File: torch_xla/experimental/custom_kernel.py [hunk 1]
```diff
@@ -222,7 +265,17 @@ def shape_dtype(q, *arg):
...
+      if segment_ids is None:
+        o = torch_xla._XLAC._xla_tpu_custom_call([q, k, v], payload, shapes,
+                                                 dtypes)
+
+      else:
+        o = torch_xla._XLAC._xla_tpu_custom_call(
+            [q, k, v, q_segment_ids, kv_segment_ids], payload, shapes, dtypes)
+
       if not save_residuals:
+        o = o[0]
```
Comment (JackCaoG) [inline torch_xla/experimental/custom_kernel.py:278]: what's this for?

File: torch_xla/experimental/custom_kernel.py [hunk 2]
```diff
@@ -222,7 +265,17 @@ def shape_dtype(q, *arg):
           min(FlashAttention.DEFAULT_BLOCK_SIZES["block_k"], k.shape[2]),
           False,
           static_argnums=range(5, 13))
+
+      if segment_ids is None:
+        o = torch_xla._XLAC._xla_tpu_custom_call([q, k, v], payload, shapes,
+                                                 dtypes)
+
```
Comment (JackCaoG) [inline torch_xla/experimental/custom_kernel.py:272]: remove empty line

File: torch_xla/experimental/custom_kernel.py [hunk 3]
```diff
@@ -357,18 +424,22 @@ def backward(ctx, grad_output):
...
+    return grad_q, grad_k, grad_v, None, None, None, None, None
 
 
 def flash_attention(
     q,  # [batch_size, num_heads, q_seq_len, d_model]
     k,  # [batch_size, num_heads, kv_seq_len, d_model]
     v,  # [batch_size, num_heads, kv_seq_len, d_model]
     causal=False,
+    q_segment_ids=None,
+    kv_segment_ids=None,
```
Comment (JackCaoG) [inline torch_xla/experimental/custom_kernel.py:436]: do we need to also update `flash_attention_xla` custom op?

**Comment by alanwaketan** [comment]
> Is this ready for review?

I still need to add spmd and dynamo support. So not yet.

File: torch_xla/experimental/custom_kernel.py (hunk 2 above)
Comment (alanwaketan) [reply torch_xla/experimental/custom_kernel.py:272]: Done.

File: torch_xla/experimental/custom_kernel.py (hunk 1 above)
Comment (alanwaketan) [reply torch_xla/experimental/custom_kernel.py:278]: _xla_tpu_custom_call always return an array.

**Comment by alanwaketan** [comment]
@JackCaoG Do you think we can do the SPMD and dynamo parts later since the customer is not using either of them now?

**Comment by JackCaoG** [comment]
yea.. don't worry about SPMD and dynamo for this pr, let's do that in a separate pr..

File: test/test_pallas.py [hunk 4]
```diff
@@ -619,6 +633,108 @@ def paged_attention_wrapper(q, k, v, seq_lens, page_indices,
...
+                jax_k,
+                jax_v,
+                segment_ids=SegmentIds(jax_q_segment_ids, jax_kv_segment_ids),
+            )))
+
+    self.assertTrue(torch.allclose(o.cpu(), expected_o.cpu(), atol=1e-05))
+
+  @unittest.skipIf(xr.device_type() != 'TPU' or tpu.version() < 3,
+                   "This test only works on TPUv3+.")
+  def test_flash_attention_wrapper_segment_ids_2(self):
```
Comment (JackCaoG) [inline test/test_pallas.py:669]: so you have 2 test, one compare to native torch, one compare to jax?

File: test/test_pallas.py [hunk 5]
```diff
@@ -619,6 +633,108 @@ def paged_attention_wrapper(q, k, v, seq_lens, page_indices,
...
+
+    torch.manual_seed(42)
+    q = torch.randn(4, 2, 128, 8, requires_grad=True).to("xla")
+    k = torch.randn(4, 2, 128, 8, requires_grad=True).to("xla")
+    v = torch.randn(4, 2, 128, 8, requires_grad=True).to("xla")
+    q_segment_ids = torch.zeros(4, 128).to("xla")
+    kv_segment_ids = torch.zeros(4, 128).to("xla")
+    q.retain_grad()
+    k.retain_grad()
+    v.retain_grad()
```
[the remark this answers is not in the document]
Comment (JackCaoG) [reply test/test_pallas.py:703]: actually just refactor this part out and uses it on all tests, it is the same for all tests.

File: torch_xla/experimental/custom_kernel.py [hunk 6]
```diff
@@ -357,18 +418,22 @@ def backward(ctx, grad_output):
       grad_v = xs.disable_manual_sharding(
           grad_v, partition_spec, full_shape, mesh=mesh).global_tensor
 
-    return grad_q, grad_k, grad_v, None, None, None
+    return grad_q, grad_k, grad_v, None, None, None, None, None
```
Comment (JackCaoG) [inline torch_xla/experimental/custom_kernel.py:421]: why do we need to return these Nones?

File: test/test_pallas.py (hunk 4 above)
Comment (alanwaketan) [reply test/test_pallas.py:669]: Yea, the JAX test is written before I figure out how to do the non-kernel mask.

File: test/test_pallas.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alanwaketan) [reply test/test_pallas.py:703]: You mean the tensor initializations? Those are kinda of expected paperworks. I don't think it's necessary to improve...

File: torch_xla/experimental/custom_kernel.py (hunk 6 above)
Comment (alanwaketan) [reply torch_xla/experimental/custom_kernel.py:421]: It's the rule of the autograd.Function where all the inputs passed in the forward need to have the corresponding grads. For inputs that we don't diff on, we return None.

File: test/test_pallas.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JackCaoG) [reply test/test_pallas.py:703]: I will leave that to you. When I see two large chunks of codes that looks similar, I usually tried to find how they are different. It confused me a bit when I realized it is the same code repeating over and over.

**Review by JackCaoG — APPROVED** [review]

File: test/test_pallas.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alanwaketan) [reply test/test_pallas.py:703]: Yea, for testing, it's sometime hard to avoid... haha

**Comment by alanwaketan** [comment]
Thanks, Jack.
